File: bubble-sort/typescript/bubble-sort.ts
/**
 * 冒泡排序
 * @param arr     要排序的数组
 * @param compare 用于排序比较的函数
 */
export default function bubbleSort<T>(arr: T[], compare: (a: T, b: T) => number = (a, b) => a > b ? 1 : a < b ? -1 : 0): T[] {
	// 将数值进行复制
	arr = Array.from(arr);
	// n 为数组长度
	const n = arr.length;
	for (let i = n - 1; i > 0; i--) {
		// 标记此轮是否有过交换
		let swapped = false;
		for (let j = 0; j < i; j++) {
			if (compare(arr[j], arr[j + 1]) > 0) {
				// 交换两个数值
				[arr[j], arr[j + 1]] = [arr[j + 1], arr[j]];
				// 标记此轮有交换过
				swapped = true;
			}
		}
		// 若此轮没有交换过，表示已经排好序列，直接返回
		if (!swapped) {
			break;
		}
	}
	return arr;
}

File: bubble-sort/matlab/bubble_sort.m
function arr = bubble_sort(arr)
%bubble_sort - 对数组从小到大排列
%
% Syntax: sort_arr = bubble_sort(arr)
%
% 使用冒泡排序对数组从小到大排序
	len = length(arr);
	for i = 1 : (len - 1)
		swapped = false;
		for j = 1 : (len - i)
			if arr(j) > arr(j + 1)
				temp = arr(j);
				arr(j) = arr(j + 1);
				arr(j + 1) = temp;
				swapped = true;
			end
		end
		if ~swapped
			break;
		end
	end
end